File: guides/micronaut-data-r2dbc-repository/kotlin/src/main/kotlin/example/micronaut/GenreRepository.kt
package example.micronaut

import example.micronaut.domain.Genre
import io.micronaut.data.annotation.Id
import io.micronaut.data.exceptions.DataAccessException
import io.micronaut.data.model.query.builder.sql.Dialect
import io.micronaut.data.r2dbc.annotation.R2dbcRepository
import io.micronaut.data.repository.reactive.ReactorPageableRepository
import reactor.core.publisher.Mono
import jakarta.transaction.Transactional
import jakarta.validation.constraints.NotBlank

@R2dbcRepository(dialect = Dialect.MYSQL) // <1>
abstract class GenreRepository : ReactorPageableRepository<Genre, Long> { // <2>
    abstract fun save(@NotBlank name: String): Mono<Genre>

    @Transactional
    open fun saveWithException(@NotBlank name: String): Mono<Genre> {
        return save(name)
            .then(Mono.error(DataAccessException("test exception")))
    }

    abstract fun update(@Id id: Long, @NotBlank name: String): Mono<Long>
}

File: guides/micronaut-data-r2dbc-repository/python/src/example/micronaut/genre_repository.py
from typing import Annotated

import java
from jakarta.transaction import Transactional
from jakarta.validation.constraints import NotBlank
from micronaut.data.annotation import Id
from micronaut.data.exceptions import DataAccessException
from micronaut.data.model.query.builder.sql import Dialect
from micronaut.data.r2dbc.annotation import R2dbcRepository
from micronaut.data.repository.reactive import ReactorPageableRepository
from org.reactivestreams import Publisher
from reactor.core.publisher import Mono

from .domain.genre import Genre


@R2dbcRepository(dialect=Dialect.MYSQL)  # <1>
class GenreRepository(ReactorPageableRepository[Genre, int]):  # <2>
    def save(self, name: Annotated[str, NotBlank]) -> Publisher[Genre]: ...

    @Transactional
    def saveWithException(
        self,
        name: Annotated[str, NotBlank],
    ) -> Publisher[Genre]:
        return Mono.from_(self.save(name)).then(Mono.error(DataAccessException("test exception")))

    def update(
        self,
        id: Annotated[int, Id],
        name: Annotated[str, NotBlank],
    ) -> Publisher[int]: ...
